dask_mpi/tests/basic.py

```diff
@@ -0,0 +1,27 @@
+import sys
+
+from time import sleep
+from distributed import Client
+from distributed.metrics import time
+from tornado import gen
+
+from dask_mpi.core import initialize
+
+scheduler_file = sys.argv[1]
+
+initialize(scheduler_file=scheduler_file)
+
+with Client(scheduler_file=scheduler_file) as c:
+    start = time()
+    while len(c.scheduler_info()['workers']) != 2:
+        assert time() < start + 10
+        sleep(0.2)
+
+    assert c.submit(lambda x: x + 1, 10, workers=1).result() == 11
+
+    async def stop(dask_scheduler):
+        await dask_scheduler.close()
+        await gen.sleep(0.1)
+        loop = dask_scheduler.loop
+        loop.add_callback(loop.stop)
+    c.run_on_scheduler(stop, wait=False)
```

Comment:
We might want to add this somehow to the initialize script so that the user doesn't have to do it explicitly.  

I wonder if this would be a good match for the `atexit` module or if maybe initialize provides a client to the user and that we add a `finalize` call to that client.

I can help with this, more in a bit, have to run to a meeting.

Reply:
Yes.  I agree.  And I was actually thinking of using the `atexit` module for this, but I didn't get around to it.

Reply:
This could also be handled by an `MPICluster` object.

Reply:
:+1 on not having to do this explcitly.

Reply:
OK, so I suggest that we try something like the following:

```python
import atexit  # a python module to run code just before the process exits

@atexit.register
def send_close_signal():
    async def stop(dask_scheduler):
        await dask_scheduler.close()
        await gen.sleep(0.1)
        loop = dask_scheduler.loop
        loop.add_callback(loop.stop)

    with Client() as c:  # assuming that we've already placed the `scheduler_address` in `dask.config`
        c.run_on_scheduler(stop, wait=False)
```

I haven't actually tested this, but I *think* that it should work.

Reply:
Excellent!  I'll try it out and test it and let you know.

Reply:
Sadly, this doesn't seem to work.  I get this message in the logs:

```
distributed.protocol.pickle - INFO - Failed to serialize <function send_close_signal.<locals>.stop at 0x10d3f47b8>. Exception: can't pickle select.kqueue objects
```

Reply:
Can you verify again that you're still on dask/distributed master?

Maybe try moving the `stop` function outside of the `send_close_signal` function.  Having it as a top-level function might improve serializability.

Reply:
Ack!  You are correct.  I didn't realize this was new code in `master`.  I did a git-pull and it works.

Reply:
I've moved this code into `core.py` with the one minor exception, that I need to check for `rank == 1`: before registering the call.  So, the code block that works is:

```python
import atexit
from mpi4py import MPI

comm = MPI.COMM_WORLD
rank = comm.Get_rank()

def send_close_signal():
    async def stop(dask_scheduler):
        await dask_scheduler.close()
        await gen.sleep(0.1)
        loop = dask_scheduler.loop
        loop.add_callback(loop.stop)

    with Client() as c:  # assuming that we've already placed the `scheduler_address` in `dask.config`
        c.run_on_scheduler(stop, wait=False)

if rank == 1:
    atexit.register(send_close_signal)
```